Probes ipc4 support

jsarha opened the pull request:
IPC4 probes support. There are some features missing and the debugfs API should need some improvement for presenting IPC4 connection points better, but with this PR it is possible to extract at least the logs from IPC4 probes interface and there is no reason to believe that audio stream probes would not work (as long as you know how to identify a probe point).

plbossart reviewed (changes requested):
ERROR: Missing Signed-off-by: line(s)

jsarha commented:
The commit was  not ready and I did not want to put my signed-of for it yet. But now it should be in better shape.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc3.c:
```diff
@@ -0,0 +1,226 @@
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@intel.com>
```
I would have probably kept the original author, this is just moving the code.

jsarha replied:
Sure.

ujfalusi commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -34,19 +34,9 @@ struct sof_probes_priv {
-	unsigned int dma_buffer_size;
-} __packed;
-
 struct sof_ipc_probe_dma_add_params {
```
Do you need to keep any of the ipc3 structures in the generic code?

jsarha replied:
Only those that are needed in the probes ipc ops API. Possibly the API will be changed to use more generic structs, in which case the ipc3 structs can be local to ipc3 file.

ujfalusi replied:
But this struct is now only used in the ipc3 file, in `ipc3_probes_init()`
The `sof_ipc_probe_info_params`, `sof_ipc_probe_point_add_params` and `sof_ipc_probe_point_remove_params` also ipc3 only.

jsarha replied:
Thanks, and that is not the only struct I left lying around. Will also remove all ipc3 msg structs from sof-client-probes.c as they are not needed there anymore, as they are already in sof-client-probes-ipc3.c.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,289 @@
+static struct sof_man4_module *sof_ipc4_probe_get_module_info(struct sof_client_dev *cdev)
+{
+	struct sof_probes_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = cdev->sdev;
```
Inside sof-client driver you must _not_ access `sdev`, it is not allowed.

jsarha replied:
Is there then some API or something to look up the probes module info without directly accessing sdev?

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,289 @@
+	msg.data_size = sizeof(cfg);
+	msg.data_ptr = &cfg;
+
+	ret = iops->tx_msg(sdev, &msg, msg.data_size, NULL, 0, false);
```
No, use the `sof_client_ipc_tx_message()`

jsarha replied:
This is fine, in this place, but replacing  set_get_data() calls with sof_client_ipc_tx_message() is not that straight forwards. At least in theory the payload may not fit into a single message, and even worse I do not know how to check for that without accessing the sdev?  What about if I add the set_get_data into the sof client API?

ujfalusi replied:
If the payload can not fit into a single payload than you can only use SOF_IPC4_MOD_LARGE_CONFIG_SET message which is completely different than the SOF_IPC4_MOD_INIT_INSTANCE.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,289 @@
+			dev_err(sdev->dev, "%s: no matching uuid found", __func__);
+	}
+
+	return (struct sof_man4_module *) priv->ipc_priv;
```
I think you will need to rework this after https://github.com/thesofproject/linux/pull/3826 is merged and have a sof-client wrapper to get a module by uuid..

ujfalusi commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -531,7 +517,12 @@ static int sof_probes_client_probe(struct auxiliary_device *auxdev,
+	if (sof_client_get_ipc_type(cdev) == SOF_IPC)
+		priv->ipc_ops = &ipc3_probe_ops;
+	else
+		priv->ipc_ops = &ipc4_probe_ops;
```
and if neither you will have NULL pointer dereference....

ujfalusi commented on sound/soc/sof/sof-client-probes.h:
```diff
@@ -54,4 +54,17 @@ struct sof_probes_ipc_ops {
+	struct dentry *dfs_points_remove;
+	u32 extractor_stream_tag;
+	struct snd_soc_card card;
+	void *ipc_priv;
```
it would be nice to document now these members.

jsarha replied:
It sure would, but I am only the mover guy :). I try to extract the information from some where.

ujfalusi reviewed (changes requested):
@jsarha, It looks almost good with the exception of accessing the forbidden `sdev` and some smaller issues I can see at first scan.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	msg.primary = mentry->id;
+	msg.primary |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
```
you don't need to set the type here, it is done in the core.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+	msg.data_size = sizeof(*points) * num_desc;
+	msg.data_ptr = points;
+
+	ret = sof_client_ipc_set_get_data(cdev, &msg, true);
```
:thumbsup:

jsarha commented:
This PR is now rebased on top of: https://github.com/thesofproject/linux/pull/3826 so only the last 5 commits are relevant to this PR. The rest are coming from #3826.

ujfalusi commented on sound/soc/sof/sof-client.c:
```diff
@@ -284,6 +285,20 @@ int sof_client_ipc_set_get_data(struct sof_client_dev *cdev, void *ipc_msg,
+	if (sdev->pdata->ipc_type == SOF_INTEL_IPC4)
+		return sof_ipc4_find_module_by_uuid(sdev, uuid);
+#endif
+	dev_err(sdev->dev, "%s: IPC4 required", __func__);
```
I guess if IPC4 is not enabled the ipc4 part of the probes will not going to be built also?

I would cover the whole internals of the function except the `return NULL;` and change the dev_err print to "Only supported with IPC4\n"

jsarha replied:
I think whether IPC4 support is built and whether its in use are two different things. The ifdef is there to avoid undefined symbol if IPC4 support is not built. If statement maybe unnecessary if sof_ipc4_find_module_by_uuid() behaves well even if IPC4 is not in use, but to me it does not look like that. Will update the error print.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+	struct {
+		u32 vindex:8;
+		u32 dma_type:5;
+		u32 rsvd:19;
```
I usually try to avoid this type of representation and use shifts, masks.
In here it is not that hard to know  that dma_type is GENMASK(12, 8), but with more fields it gets into the way.

jsarha replied:
That struct was copied from avs driver, and sof fw uses the same representation. I am actually not sure in which order the bit-field bits end up in the integer, so I avoided making the conversion. If you insist, I do it by trial and error, but you must insist 😅.

ranj063 replied:
for the sake of consistency in the SOF driver, I insist @jsarha :)

jsarha replied:
Ok, I will do it then. However, I still think it does not help with the code readability, and the probe_point_id starts to look dangerously similar to IPC3 buffer_id.

jsarha replied:
Or can I interpret this so that the "union sof_ipc4_probe_point_id" if fine as it is, but for DMA I should use the already existing macros for IPC4 node_id (Sorry I found them only now.).... Well I try that 😅.

ujfalusi replied:
@jsarha, what do you mean (and found) regarding to `DMA I should use the already existing macros for IPC4 node_id`?
_That_ `node_id` in `struct sof_copier_gateway_cfg` is a completely different (for convenience) than the `node_id` for the probes's `struct sof_ipc4_probe_gtw_cfg`

I think what would be nicer to do is to get rid of the union, use u32 and macros to shift things into place as we do in other places in SOF kernel code.

jsarha replied:
Ok, its not at all trivial to draw the connections between payload struct data members and macros that should be used to fill the bit-fields within those members. Not at all convinced this is better system. I was fooled that a node_id is a node_id,  and same macros can be used anywhere since the naming and bit-fields match.  Well,  I try to do better job with IPC4 PROBES macros.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+{
+	struct device *dev = &cdev->auxdev.dev;
+	struct sof_man4_module *mentry =
+		sof_ipc4_probe_get_module_info(cdev);
```
this can fit to one line.
also in other functions.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+	int ret;
+
+	if (!mentry)
+		return -EINVAL;
```
It is more likely -ENODEV or -EOPNOTSUPP since the probes module itself is not part of the firmware.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+	int ret;
+
+	if (!mentry)
+		return -EINVAL;
```
This can not happen if we are already here, the `ipc4_probes_init()` would have failed.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+
+	ret = sof_client_ipc_tx_message(cdev, &msg, NULL, 0);
+	if (ret)
+		dev_err(dev, "%s() failed with: %d", __func__, ret);
```
I think we are going to have enough prints from the IPC layer and then from the generic probes code, can be dropped?

jsarha replied:
True, I'll drop all the error prints from ipc_ops.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,276 @@
+
+	ret = sof_client_ipc_set_get_data(cdev, &msg, true);
+	if (ret)
+		dev_err(dev, "%s() failed with: %d", __func__, ret);
```
Same here, I think we are going to be aware of the error codes from other layers.

ujfalusi reviewed (commented):
@jsarha, this looks really nice, I have only few comments, most of them are nit-picks.

jsarha commented:
Fixed comments from Peter, checkpatch issues and compiler warnings.

plbossart commented:
@jsarha can you rebase and resubmit? the initial dependencies were merged, this should make your PR much easier to review.

jsarha commented:
> @jsarha can you rebase and resubmit? the initial dependencies were merged, this should make your PR much easier to review.

Sure. Done. Also tested that it still works.

plbossart commented on sound/soc/sof/sof-client-probes-ipc3.c:
```diff
@@ -0,0 +1,235 @@
+	unsigned int num_elems;
+	union {
+		struct sof_probe_dma dma[0];
+		struct sof_probe_point_desc desc[0];
```
aren't we supposed to use [] now? This should be fixed before moving the code to a new file, no?

jsarha replied:
This is simple move code. I would not mix even a simple syntax update like that with this commit. If we want to change [0] to [] I would go with a single commit fixing all sof source.

plbossart replied:
these are the only two places where array[0] is used instead of array[]. Easy to fix and then move the code.

jsarha replied:
Ok, now I know why these four places still have [0]. They are all within union and:

/home/oku/ktest/sof-dev/sound/soc/sof/sof-client-probes-ipc3.c:31:24: error: flexible array member in union
   31 |   struct sof_probe_dma dma[];
      |                        ^~~

plbossart replied:
Thanks for checking @jsarha. Indeed the transition to flexible arrays is only for struct, not union, for some reason.

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -284,6 +285,20 @@ int sof_client_ipc_set_get_data(struct sof_client_dev *cdev, void *ipc_msg,
+#endif
+	dev_err(sdev->dev, "Only supported with IPC4\n");
+
+	return NULL;
```
this is rather odd, do we actually have a dependency on a Kconfig IPC4 setting when this is only relevant for IPC4. Something's weird, can we simplify?

jsarha replied:
We have a Kconfig options for compiling in IPC3 support, IPC4 support, or both, then we have a private data variable telling which one should be used.  So its possible IPC4 not built in, in which case the #if is necessary to avoid undefined symbol. But its also possible that IPC4 is built, but it just is not in use (depends on the module parameters). Then again the sof_client API is common for both IPC modes, but this particular function does not make any sense in IPC3 mode. I do not know any better way to deal with the situation.

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -284,6 +285,20 @@ int sof_client_ipc_set_get_data(struct sof_client_dev *cdev, void *ipc_msg,
+	if (sdev->pdata->ipc_type == SOF_INTEL_IPC4)
+		return sof_ipc4_find_module_by_uuid(sdev, uuid);
+#endif
+	dev_err(sdev->dev, "Only supported with IPC4\n");
```
why would an ipc4-related code throw an error 'only supported with IPC4'? not getting this at all.

jsarha replied:
I can of course drop the error message if that is the problem. It is there only to help with finding out programming errors sog-client calling ipc4 only functions in ipc3 mode. The problem arises from having to add IPC4 specific function to generic sof-client API.

plbossart replied:
well, this error cannot happen, ipc4_probes_init is completely tied to IPC4 already. see next patch.

ujfalusi replied:
imho the whole `sof_client_ipc4_find_module_by_uuid()` can be within the #ifdef , any code which might use it should be only compiled if ipc4 support is compiled.
If this is not going to be the case then provide an static inline stub in the header with the same condition?

ranj063 replied:
why not call this function sof_client_find_module() if you want to keep the test for IPC type and ensure you can do this only for IPC4? That way, maybe we don';t even need the ifdef around the function

ujfalusi replied:
> why not call this function sof_client_find_module() if you want to keep the test for IPC type and ensure you can do this only for IPC4? That way, maybe we don';t even need the ifdef around the function

We can only find a module with IPC4 and we only have ipc4 private support for this.

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+
+	ret = sof_client_ipc_tx_message(cdev, &msg, NULL, 0);
+
+	return ret;
```
return sof_client_ipc_tx_message(cdev, &msg, NULL, 0); and remove ret?

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+
+struct sof_ipc4_probe_point {
+	union sof_ipc4_probe_point_id buffer_id;
+	u32 purpose;
```
what is the definition of purpose, or the purpose of this definition?

jsarha replied:
It depends whether we are in IPC3 [1] or IPC4 [2] mode 😅. In any case the current Linux driver probes implementation tries to be agnostic about the FW details and expose bare numbers though debugfs. The debugfs is then in turn used by user space tools compiled from sof FW sources, sharing the same headers. This is after all debug interface for developers.

Hmm, maybe we should start purpose counting from 0 on IPC3 side too, since the IPC3 API has hardly been ever used outside SOF team.

[1] https://github.com/thesofproject/sof/blob/1bca97ad4cf4e8eb45bd4ce6ca3ac87b499914ee/src/include/ipc3/probe.h#L29

[2] https://github.com/thesofproject/sof/blob/1bca97ad4cf4e8eb45bd4ce6ca3ac87b499914ee/src/include/ipc4/probe.h#L38

ujfalusi replied:
This is IPC4 probes code, so what IPC3 defines is irrelevant? If you are keeping the debugfs interface generic, then either IPC3 or IPC4 must align with the core's convention.

kv2019i replied:
@ujfalusi @plbossart The probe points are uint32 identifiers. Their allocation policy is different in IPC3 and IPC4, but the current debugfs interface is agnostic to this. In both IPC3 and IPC4, user of debugfs has to have some out-of-band means to figure out which IDs are correct, so no change here because of IPC4.

I do think we need to consider a more user-friendly debugfs interface in the future, but I think we can do that on top. E.g. add introspection (you can query which probe points are available in the system) and then a more user-friendly (and less error prone way) to set them. But I'd say this is beyond this PR (as situation was not much better with IPC3).

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+	struct {
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
```
what does type refer to?

ujfalusi replied:
My guess (after reading https://github.com/thesofproject/sof/blob/main/src/include/ipc4/probe.h) it is either input/output/internal. Not that this information helps...

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+};
+
+union sof_ipc4_probe_point_id {
+	u32 full_id;
```
what is that?

jsarha replied:
See bellow.

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
+		u32 index:6;
```
confusing, there's full_id and index, what is the difference or overlap?

jsarha replied:
To be honest I do not know exactly what all these bit-fields mean exactly. The struct was copied from firmware side, and the same struct is apparently used by windows driver too. Somehow the struct members identify the audio buffer to probe. I had to use fw-side prints to find some number that were accepted by the firmware and topology combination. On IPC3 side the buffer_id alone is used for the same purpose.

ujfalusi replied:
I agree that the naming is confusing (which is not that uncommon within IPC4)...
I would rename the `full_id` to `val` or `val32`. This is used within `sof_ipc4_probe_point` as buffer_id - which is misleading also as it is not really a buffer ID but a composite of assorted parameters describing a buffer...

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,249 @@
+
+	ret = sof_client_ipc_tx_message(cdev, &msg, NULL, 0);
+
+	return ret;
```
return sof_client_ipc_tx_message(cdev, &msg, NULL, 0); ?

plbossart commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -430,7 +416,22 @@ static int sof_probes_client_probe(struct auxiliary_device *auxdev,
-	priv->ipc_ops = &ipc3_probe_ops;
+
+	if (sof_client_get_ipc_type(cdev) == SOF_INTEL_IPC4) {
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_IPC4)
```
this is also odd, if we are in type SOF_INTEL_IPC4 then the IPC4 needs to be enabled, no? can we simplify those ifdef or use IS_ENABLED() ?

jsarha replied:
<argh, editing comment aftewards and then deciding that it is good as it was end pressing cancel deletes the commtent ... writing the comment again as I remember it>

Yes for ipc type to be SOF_INTEL_IPC4, the CONFIG_SND_SOC_SOF_INTEL_IPC4 has to be enabled, but the refences to IPC4 code has to be enclosed behind #if to avoid undefined symbols, if IPC4 is not enabled. But yes simple #ifdef should do as well. Should I do the switch? Also to sof_client_ipc4_find_module_by_uuid() in sof-client.c?

plbossart commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -430,7 +416,22 @@ static int sof_probes_client_probe(struct auxiliary_device *auxdev,
+	} else if (sof_client_get_ipc_type(cdev) == SOF_IPC) {
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_IPC3)
+		priv->ipc_ops = &ipc3_probe_ops;
+#endif
```
same, SOF_IPC means IPC3. Can we simplify?

jsarha replied:
Sure. Just thought that IS_ENABLED() macro is there for a purpose and should be used when checking Kconfig options. But now I see that #ifdef CONFIG_* is 10 times more common than #if IS_ENABLED(CONFIG_*), so it should be just as good.

RanderWang commented on sound/soc/sof/sof-client.c:
```diff
@@ -265,6 +265,25 @@ int sof_client_ipc_tx_message(struct sof_client_dev *cdev, void *ipc_msg,
+
+	return -EINVAL;
+}
+EXPORT_SYMBOL_NS_GPL(sof_client_ipc_set_get_data, SND_SOC_SOF_CLIENT);
```
use "return ipc->ops->set_get_data(ipc->sdev, msg_data, msg_bytes, set); " in sof_ipc_set_get_data
why we need to check ipc_type ? change it to ?
````c
int sof_client_ipc_set_get_data(struct sof_client_dev *cdev, void *ipc_msg, int size,
				bool set)
{
		return sof_ipc_set_get_data(cdev->sdev->ipc, ipc_msg, size, set);
}
````

jsarha replied:
> use "return ipc->ops->set_get_data(ipc->sdev, msg_data, msg_bytes, set); " in sof_ipc_set_get_data

I just thought would be good practice to follow same implementation pattern as in sof_client_ipc_tx_message().

> why we need to check ipc_type ? change it to ?

The function prototype does not take message size separately and needs to cast the message pointer to a struct of the right form to take the size from message struct itself.

RanderWang commented on sound/soc/sof/sof-client.h:
```diff
@@ -39,6 +41,7 @@ int sof_client_ipc_tx_message(struct sof_client_dev *cdev, void *ipc_msg,
 int sof_client_ipc_set_get_data(struct sof_client_dev *cdev, void *ipc_msg,
 				bool set);
+struct sof_ipc4_fw_module *sof_client_ipc4_find_module_by_uuid(struct sof_client_dev *cdev, const guid_t *uuid);
 
```
this function is only used by ipc4-probe.c, we don't need to define it here for general client.
If you want to make it useable for future,  I recommend it define in ipc4-loader.c and add anther ops in struct sof_ipc_fw_loader_ops

jsarha replied:
I understood from earlier comments that a sof-client should not access struct snd_sof_dev directly, and if that stands a function using client API is needed.

ujfalusi replied:
Yes, that's right. SOF clients _must_ not access anything from the core directly and the access should be kept minimum.

jsarha commented:
Applied @plbossart 's suggestions.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,243 @@
+{
+	struct sof_man4_module *mentry = sof_ipc4_probe_get_module_info(cdev);
+	struct sof_ipc4_msg msg;
+	struct sof_ipc4_probe_point *points;
```
For a nice descending end of line, swap the two lines (and also in other functions).

ujfalusi commented on sound/soc/sof/sof-client.c:
```diff
@@ -289,7 +289,7 @@ struct sof_ipc4_fw_module *sof_client_ipc4_find_module_by_uuid(struct sof_client
 	struct snd_sof_dev *sdev = cdev->sdev;
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_IPC4)
+#ifdef CONFIG_SND_SOC_SOF_INTEL_IPC4
```
Move this when `sof_client_ipc4_find_module_by_uuid()` was introduced?

plbossart replied:
use if(IS_ENABLED), it preserves indentation and is nicer to read.

jsarha replied:
> Move this when `sof_client_ipc4_find_module_by_uuid()` was introduced?

Yep, will do, and I'll put the whole function behind ifdefs as you suggested. No sense to define a function if its not supported.

ujfalusi reviewed (commented):
@jsarha, Only nit-picking...

jsarha commented:
Hope this is now good...

plbossart commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -430,7 +416,22 @@ static int sof_probes_client_probe(struct auxiliary_device *auxdev,
-	priv->ipc_ops = &ipc3_probe_ops;
+
+	if (sof_client_get_ipc_type(cdev) == SOF_INTEL_IPC4) {
+#ifdef CONFIG_SND_SOC_SOF_INTEL_IPC4
```
nit-pick: i prefer the use of if (IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL_IPC4)) to avoid breaking the indentation. The compiler is smart-enough to prune always-false branches.

ranj063 replied:
can we change the if/ifelse to a switch case? will make it easier to add new IPCs later on

jsarha replied:
Sure, I can use switch(sof_client_get_ipc_type(cdev)), but the references to IPC4 code must be filtered out by preprocessor  to avoid "undefined symbols".

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -152,6 +184,15 @@ static int ipc4_probes_deinit(struct sof_client_dev *cdev)
+ * @cdev:	SOF client device
+ * @desc:	Returned list of active probes
+ * @num_desc:	Returned count of active probes
+ * @return:		0 on success, negative error code on error
```
not sure why we have an indentation issue here?

jsarha replied:
Me neither. Will fix.

plbossart commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -162,6 +203,16 @@ static int ipc4_probes_points_info(struct sof_client_dev *cdev,
+ * @cdev:	SOF client device
+ * @desc:	List of probe points to connect
+ * @num_desc:	Number of elements in @desc
+ * @return:		0 on success, negative error code on error
```
not sure why we have an indentation issue here?

plbossart reviewed (approved):
Looks good-enough to me, only nit-picks at this point. Let's try to merge this ASAP. Thanks @jsarha

ranj063 commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,243 @@
+	struct device *dev = &cdev->auxdev.dev;
+	static const guid_t probe_uuid =
+		GUID_INIT(0x7CAD0808, 0xAB10, 0xCD23,
+			  0xEF, 0x45, 0x12, 0xAB, 0x34, 0xCD, 0x56, 0xEF);
```
@jsarha for my education, can you please explain what this UUID is for? Do we need to load an external library to support probes for IPC4?

kv2019i replied:
@ranj063 Probes client will be built-in, but this is just practical setup. With loadable module support merged, probes could be moved to a separate module later on.

ranj063 commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,243 @@
+	};
+
+	if (!mentry)
+		return -ENODEV;
```
ENODEV or ENOENT?

ranj063 commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -80,6 +80,14 @@ struct sof_ipc4_probe_point {
 
 #define INVALID_PIPELINE_ID      0xFF
 
+/**
```
can we squash this with the previous patch that introduces the IP4 ops?

jsarha replied:
Sure.

ujfalusi commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,304 @@
+struct sof_ipc4_probe_dma {
+	union sof_ipc4_connector_node_id node_id;
+	unsigned int dma_buffer_size;
+} __packed __aligned(4);
```
This struct is unused?

jsarha replied:
Yes, somehow missed it when I removed all the other structs that after all were not needed.

jsarha commented:
One more round. Dropped one unused struct and two bit-field unions (another in favor of corresponding field setting macroes, the other was not needed at all), squashed the IPC4 doxygen documentation commit together with IPC4 code commit, used switch instead of if-else in the init code, couple of indentation fixes, and one idle include-line removed.

jsarha commented:
... and fix ipc3_probes_init -> ipc4_probes_init in doxygen.

jsarha commented:
... and space after switch ().

kv2019i commented on sound/soc/sof/sof-client-probes-ipc4.c:
```diff
@@ -0,0 +1,281 @@
+				   struct sof_probe_point_desc **desc,
+				   size_t *num_desc)
+{
+	/* TODO: Firmware side implementation needed first */
```
Not for this PR, but I think this introspection could be queried from the kernel driver's information on the topology. No need to go and ask FW about the probe points, kernel already knows when it has parsed the topology. It also knows what is loaded and so forth.

jsarha replied:
Absolutely. I have actually been thinking that it would be nice if, like the probe_points  debugfs file - when read -  lists currently active probe points , the probe_points_remove could list all the available probe points in the system with as much information that is available to the firmware.

I was also thinking of putting in a change in the firmware that turns on the log probe as soon as probe dma is started. This is anyway what @lgirdwood asked me to implement in the probes tool, but its just much simpler to by pass the API agnostic debugfs interface and do it in the FW.

kv2019i reviewed (approved):
Thanks @jsarha , I'm good with this (please do address comments already added by others). I think the general question of how to make the probe-point interface more user-friendly, needs to be addressed, but this is an old problem (present in the old IPC3 interface as well), so I'd not block this PR and start trying to solve that here.

plbossart commented on sound/soc/sof/sof-client.h:
```diff
@@ -39,7 +41,9 @@ int sof_client_ipc_tx_message(struct sof_client_dev *cdev, void *ipc_msg,
-
+#ifdef CONFIG_SND_SOC_SOF_INTEL_IPC4
+struct sof_ipc4_fw_module *sof_client_ipc4_find_module_by_uuid(struct sof_client_dev *cdev, const guid_t *uuid);
+#endif
```
this #ifdfef seems unnecessary in a header file, and inconsistent with the inclusion of 'struct sof_ipc4_fw_module' unconditionally above.

jsarha replied:
Indeed, declaration alone does not cause reference. Thanks, I'll fix that.

plbossart reviewed (dismissed):
only one nit-pick below, thanks @jsarha

ujfalusi reviewed (approved):
@jsarha, from my side, this looks nice.
Thank you!

Can you address the two fixable  checkpatch issue:
```
WARNING: line length of 111 exceeds 100 columns
#33: FILE: sound/soc/sof/sof-client.c:289:
+struct sof_ipc4_fw_module *sof_client_ipc4_find_module_by_uuid(struct sof_client_dev *cdev, const guid_t *uuid)

WARNING: line length of 112 exceeds 100 columns
#66: FILE: sound/soc/sof/sof-client.h:45:
+struct sof_ipc4_fw_module *sof_client_ipc4_find_module_by_uuid(struct sof_client_dev *cdev, const guid_t *uuid);
```

jsarha commented:
Also the long line complains by checkpatch should now be fixed.

jsarha commented:
One more commit, this time the beginning of the series. That should at least require reviewing still, the other changes are only the DECLARE_FLEX_ARRAY()-usage also in sof-client-probes-ipc3.c in the second commit. Otherwise this is identical to all earlier versions. I also tested that probes and probe-logs work with both ipc3 and ipc4 setup.

jsarha commented:
... fixed couple of typos in the commit message of the first commit.

ujfalusi commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -206,8 +207,8 @@ struct sof_ipc4_control_data {
-		struct sof_ipc4_ctrl_value_chan chanv[0];
-		struct sof_abi_hdr data[0];
+		DECLARE_FLEX_ARRAY(struct sof_ipc4_ctrl_value_chan, chanv);
+		DECLARE_FLEX_ARRAY(struct sof_abi_hdr, data);
```
Before:
```
sound/soc/sof/ipc4-control.c: note: in included file:
sound/soc/sof/ipc4-topology.h:210:40: warning: array of flexible structures

sound/soc/sof/ipc4-topology.c: note: in included file:
sound/soc/sof/ipc4-topology.h:210:40: warning: array of flexible structures
```

After:
```
sound/soc/sof/ipc4-control.c: note: in included file:
sound/soc/sof/ipc4-topology.h:211:17: warning: array of flexible structures

sound/soc/sof/ipc4-topology.c: note: in included file:
sound/soc/sof/ipc4-topology.h:211:17: warning: array of flexible structures
```
The problem is that `struct sof_abi_hdr` itself has flexible array (`__u32 data[];`), a flexible array of a flexible struct is hard to digest..

jsarha replied:
Ok... just brainlessly applied the macros in place of [0]. My first reaction would be to change struct sof_abi_hdr data[0]; in struct sof_ipc4_control_data to simply struct sof_abi_hdr data; ... but I do not know the code that well. In any case flex array of flex arrays must be a bug, so one of the flex arrays must go.

jsarha replied:
Well, it seems I can change "struct sof_abi_hdr data[0];" to "struct sof_abi_hdr data;" in both places without any changes anywhere... so it looks like those data members are not touched anywhere. But I am completely lost here, I have no idea what these structs are for and why they have been made flexible in the first place.... but my IPC4 probes changes has nothing to do with those and should get merged soon, so I drop everything else but the probes related changes.

ujfalusi replied:
Some reason Intel related things have this love for flexible array of flexible arrays nested.

Imho it is just plain wrong to use a struct (which is a flexible struct) as a flexible member...

```c
struct sof_ipc4_control_data {
	struct sof_ipc4_msg msg;
	int index;

	union {
		struct sof_ipc4_ctrl_value_chan chanv[0];
		struct sof_abi_hdr data[0];
	};
} blah;
```
Where `blah.data[2]` is pointing?

My uneducated fixes for similar issues for NHLT: 62c03bcc4ad3b529f1d954ad705557f53c566b78, 2b79c659507d8dd00d0fe0cbec95290fd23dfe4b and one for SOF: 490217ea34e4b33566c7d360959f24b32c6c6e95.
I'm not sure if this is correct, but I don't see other way to fix it.

I can not just drop the `[]` from the end as we might have multiple flexible structs placed after each other, but having it named as a struct is not right either.

ujfalusi commented on include/sound/sof/control.h:
```diff
@@ -149,7 +149,7 @@ struct sof_ipc_comp_event {
 	union {
 		/* data can be used by binary controls */
-		struct sof_abi_hdr data[0];
+		DECLARE_FLEX_ARRAY(struct sof_abi_hdr, data);
```
Sparse stops earlier, but this will not help either...
```
sound/soc/sof/ipc3-topology.c: note: in included file (through sound/soc/sof/sof-audio.h):
include/sound/sof/control.h:124:40: warning: array of flexible structures
```

jsarha commented:
Dropped all non essential [0] to DECLARE_FLEX_ARRAY() changes, to get this PR forwards. Happy to fix the remaining [0] issues, but not within this PR.

plbossart commented:
> Dropped all non essential [0] to DECLARE_FLEX_ARRAY() changes, to get this PR forwards. Happy to fix the remaining [0] issues, but not within this PR.

we're going to have to fix this for upstream, so this is just the same amount of work if we fix it now..

ujfalusi commented:
> > Dropped all non essential [0] to DECLARE_FLEX_ARRAY() changes, to get this PR forwards. Happy to fix the remaining [0] issues, but not within this PR.
> 
> we're going to have to fix this for upstream, so this is just the same amount of work if we fix it now..

@plbossart, I agree but this is not that simple and not the scope of this PR. I have been rolling these fixes for half a year and I'm still not decided if they are ready for even a draft PR. Other fixes for the nested flexible struct array have been already sent upstream, but these are debatable cases.
imho

plbossart commented:
@jsarha the purpose of every PR is to eventually land upstream. I have no desire to carry this in the topic/sof-dev branch if there's no clear plan, sorry.

jsarha commented:
> > Dropped all non essential [0] to DECLARE_FLEX_ARRAY() changes, to get this PR forwards. Happy to fix the remaining [0] issues, but not within this PR.
> 
> we're going to have to fix this for upstream, so this is just the same amount of work if we fix it now..

If you wish, I'll make another PR right now, but I do not think the changes outside sof-client-probes.c belongs to this PR. This is about the probes IPC4 support, not about fixing all the hidden issues within sof driver.

> @jsarha the purpose of every PR is to eventually land upstream. I have no desire to carry this in the topic/sof-dev branch if there's no clear plan, sorry.

The patches here do not include a single unfixed reference to [0] arrays, so there should be nothing in the way of applying these to uptstream.

plbossart commented on sound/soc/sof/sof-client-probes.c:
```diff
@@ -59,8 +60,8 @@ struct sof_ipc_probe_info_params {
-		struct sof_probe_dma dma[0];
-		struct sof_probe_point_desc desc[0];
+		DECLARE_FLEX_ARRAY(struct sof_probe_dma, dma);
+		DECLARE_FLEX_ARRAY(struct sof_probe_point_desc, desc);
```
that's the only change I was asking for, I don't get the long thread.

jsarha replied:
Yep, I myself volunteered to fix all [0] issues at once, but hit a can of worms while doing it, and backed off.

plbossart commented:
@ranj063 we want to give you credit for your review, can you re-approve and merge?

lgirdwood commented:
@jsarha what will be read if the user does NOT set the config ? The default config (if none set by user after module load) should be to emit the logs. i.e. arecord will just read the logs if run without config after module load.

jsarha commented:
> @jsarha what will be read if the user does NOT set the config ? The default config (if none set by user after module load) should be to emit the logs. i.e. arecord will just read the logs if run without config after module load.

The firmware does not accept any config before the probe dma is started. The dma is effectively "paused" (compressed dma does not move unless there is data to transfer), until the user adds probe points (or log) to the system by writing correct triplet of integers to the debugfs file. But the points can only be added after the probe dma is stareted.
